lstchain/image/cleaning.py

```diff
@@ -36,3 +50,228 @@ def apply_dynamic_cleaning(image, signal_pixels, threshold, fraction):
+                    qt95_pix_charge_dev) > mad_max
+
+        if outliers.sum() > 0:
+            removed_fraction = outliers.sum() / outliers.size
+            log.info(f'    Removed {removed_fraction:.2%} of pixels (outliers) '
+                     f'from pedestal median calculation')
+
+        # Now compute the median (for the whole camera) of the qt95's (for
+        # each pixel) of the charges in pedestal events. Use only reliable
+        # pixels for this, and exclude outliers:
+        n_valid_pixels = np.isfinite(qt95_pix_charge[reliable_pixels]).sum()
+        if n_valid_pixels < min_number_of_valid_pixels:
+            logging.warning(f'    Too few valid pixels ({n_valid_pixels}) for '
+                            f'calculation!')
+            median_ped_qt95_pix_charge.append(np.nan)
+        else:
+            median_ped_qt95_pix_charge.append(np.nanmedian(qt95_pix_charge[
+                                                               reliable_pixels &
+                                                               ~outliers]))
+    # convert to ndarray:
+    median_ped_qt95_pix_charge = np.array(median_ped_qt95_pix_charge)
+    number_of_pedestals = np.array(number_of_pedestals)
+
+    # Now compute the median for all processed subruns, which is more robust
+    # against e.g. subruns affected by car flashes. We exclude subruns
+    # which have less than half of the median statistics per subrun.
+    good_stats = number_of_pedestals > 0.5 * np.median(number_of_pedestals)
+
+    # First check if we have any valid subruns at all:
+    if np.isfinite(median_ped_qt95_pix_charge[good_stats]).sum() ==  0:
+        qped = np.nan
+        qped_dev = np.nan
+        additional_nsb_rate = np.nan
+        log.error(f'No valid computation was possible for run {run_number} with any of the processed subruns!')
+        return qped, additional_nsb_rate, None
+
+    qped = np.nanmedian(median_ped_qt95_pix_charge[good_stats])
+    not_outlier = np.zeros(len(median_ped_qt95_pix_charge), dtype='bool')
+    # Now we also remove outliers (subruns) if any:
+    qped_dev = median_abs_deviation(median_ped_qt95_pix_charge[good_stats])
+    not_outlier = (np.abs(median_ped_qt95_pix_charge - qped) /
+                   qped_dev) < mad_max
+
+    if (~good_stats).sum() > 0:
+        log.info(f'\nNumber of subruns with low statistics: '
+                 f'{(~good_stats).sum()} - removed from pedestal median '
+                 f'calculation')
+    if (~not_outlier).sum() > 0:
+        log.info(f'\nRemoved {(~not_outlier).sum()} outlier subruns '
+                 f'(out of {not_outlier.size}) from pedestal median '
+                 f'calculation')
+
+    # recompute with good-statistics and well-behaving runs:
+    qped = np.nanmedian(median_ped_qt95_pix_charge[good_stats & not_outlier])
+    log.info(f'\nNumber of subruns used in calculations: '
+             f'{(good_stats & not_outlier).sum()}')
```

Comment:
If the final number of subruns used for the calculations is 0, the script fails. It can happen for example for short runs with some "problematic" subruns (for example, it fails for run 19690 from 2024/11/09). It might be good to add a check to make sure there are enough subruns.

Reply:
Thanks, now an error is thrown if less than half of the processed subruns could be used in thr calculation.